.github/include/ci.py

```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/env python3
+
+import argparse
+import json
+import os
+import subprocess
+import sys
+from typing import List, Optional
+import random
+
+
+def run_command(
+    cmd: List[str], check: bool = True, env: Optional[dict] = None
+) -> subprocess.CompletedProcess:
+    """Run a command and return the result."""
+    print(f"Running: {' '.join(cmd)}", flush=True)
+    merged_env = os.environ.copy()
+    if env:
+        merged_env.update(env)
+    return subprocess.run(cmd, check=check, env=merged_env)
+
+
+def get_clippy_packages() -> List[str]:
+    """Get list of packages that should be linted with clippy."""
+    result = subprocess.run(
+        ["cargo", "metadata", "--format-version", "1"],
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+    metadata = json.loads(result.stdout)
+
+    clippy_packages = []
+    for pkg in metadata.get("packages", []):
+        pkg_metadata = pkg.get("metadata")
+        if pkg_metadata:
+            scx_metadata = pkg_metadata.get("scx")
+            if scx_metadata and scx_metadata.get("ci", {}).get("use_clippy") == True:
+                clippy_packages.append(pkg["name"])
+
+    return clippy_packages
+
+
+def test_binary_has_tests(binary_path: str) -> bool:
+    """Test whether a test binary contains any tests."""
+    result = subprocess.run(
+        [binary_path, "--list"],
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+
+    return result.stdout.strip() != "0 tests, 0 benchmarks"
+
+
+def run_format():
+    """Format all targets."""
+    print("Running format...", flush=True)
+
+    run_command(["cargo", "fmt"])
+    run_command(["git", "diff", "--exit-code"])
+    print("✓ Format completed successfully", flush=True)
+
+
+def run_build():
+    """Build all targets."""
+    print("Running build...", flush=True)
+
+    run_command(["cargo", "build", "--all-targets"])
+    print("✓ Build completed successfully", flush=True)
+
+
+def run_clippy():
+    """Run clippy on packages marked for CI linting."""
+    print("Running clippy...", flush=True)
+
+    clippy_packages = get_clippy_packages()
+    for package in clippy_packages:
+        run_command(["cargo", "clippy", "--no-deps", "-p", package, "--", "-Dwarnings"])
+
+    print("✓ Clippy checks passed", flush=True)
+
+
+def run_tests():
+    """Run the test suite."""
+    print("Running tests...", flush=True)
+
+    result = subprocess.run(
+        ["cargo", "test", "--no-run", "--message-format", "json"],
+        check=True,
+        stdout=subprocess.PIPE,
+        stderr=sys.stderr,
+        text=True,
+    )
+    cargo_results = [
+        json.loads(line) for line in result.stdout.splitlines() if line.strip()
+    ]
+    test_binaries = [
+        x["executable"]
+        for x in cargo_results
+        if x["reason"] == "compiler-artifact"
+        if x["profile"].get("test", False)
+        if x["executable"] is not None
+    ]
```

Comment:
Cargo doesn't have the equivalent of `ctest` that runs all tests in the build directory without building them, so this is the recommended alternative. It's a bit grim but won't break unless we update Cargo, and Cargo is pinned now so should be fine.